wayback-core/src/main/java/org/archive/wayback/authenticationcontrol/IPMatchesBooleanOperator.java

```diff
@@ -62,15 +65,69 @@ public void setAllowedRanges(List<String> allowedRanges) {
 			}
 		}
 	}
+	
+	/**		
+	 * @return null. this is a placeholder for Spring's getter/setter 		
+	 * 			examination		
+	 */
+	public List<String> getTrustedProxies() {
+		return null;
+	}
+
+	/**
+	 * @param trustedProxies parses each String IPRange provided for the proxies, adding them to
+	 * 		the list of IPRanges which must be ignored by the IP match operator
+	 */
+	public void setTrustedProxies(List<String> trustedProxies) {
+		this.trustedProxies = new ArrayList<IPRange>();
+		for (String ip : trustedProxies) {
+			IPRange range = new IPRange();
+			if (range.setRange(ip)) {
+				this.trustedProxies.add(range);
+			} else {
+				LOGGER.severe("Unable to parse range (" + ip + ")");
+			}
+		}
+	}
+
+
+	public String getClientIPFromForwardedForHeader(String forwardedForHeader){
+		ArrayList<String> forwardingIPs;
+		String ip = null;
+		Boolean containsIP = false;
+		if (forwardedForHeader.contains(",")) {
+			forwardingIPs = new ArrayList<String>(Arrays.asList(forwardedForHeader.replace(" ", "").split(",")));
+			Collections.reverse(forwardingIPs);
+			for (String forwardingIP : forwardingIPs){
+				for (IPRange range : trustedProxies) {
+					if (range.contains(forwardingIP)) {
+						containsIP = true;
+						break;
+					}
+				}
+				if (!containsIP ||
+					(containsIP && forwardingIPs.get(forwardingIPs.size() -1).equals(forwardingIP))) {
```

Comment:
Another thought, is it standard that if you are running proxies in front of your app servers, they are expected to discard any existing `X-Forwarded-For` header (not coming from a trusted source) in the first place and write their own? Thus you assume any `X-Forwarded-For` you are looking at should be okay? I was working on some proxy server stuff today and was reminded that we drop the `X-Forwarded-For` unless it comes from a specified trusted range. I am wondering if this is why the existing code uses the `X-Forwarded-For` without checking the remote address and that it is something that can continue to not be addressed in the code.

If it does need to be addressed in the code, then yeah, we could store the remote address and `X-Forwarded-For` separately (possibly the XFF in the new constant if that doesn't break anything because of the name of the existing constant) or stay with the single `REQUEST_REMOTE_ADDRESS` constant but append the remote address (comma separated) to any existing `X-Forwarded-For` before storing the value.